File: 0073-set-matrix-zeroes/0073-set-matrix-zeroes.cpp
// TC- O(m*n)
// SC - O(1)

class Solution {
private:
    bool isfirstcolzero = false;
public:
    void setZeroes(vector<vector<int>>& matrix) {
        // flag first cell of corresponding row and column
        for(int i=0; i<matrix.size(); i++){
            for(int j=0; j<matrix[0].size(); j++){
                if(matrix[i][j]==0){
                    if(j==0)
                        isfirstcolzero = true;
                    else {
                        matrix[i][0]=0;
                        matrix[0][j]=0;
                    }
                }
            }
        }

        // check for set flags in all rows and cols except first
        for(int i=1; i<matrix.size(); i++){
            for(int j=1; j<matrix[0].size(); j++){
                if(matrix[i][0]==0 or matrix[0][j]==0){
                    matrix[i][j]=0;
                }
            }
        }

        // check flag for first row before checking col to avoid override
        if(matrix[0][0]==0){
            for(int j=1; j<matrix[0].size(); j++){
                matrix[0][j]=0;
            }
        }

        if(isfirstcolzero){
            for(int i=0; i<matrix.size(); i++){
                matrix[i][0]=0;
            }
        }
    }
};

File: 0073-set-matrix-zeroes/0073-set-matrix-zeroes.py
# TC- O(mn)
# SC- O(1)

class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        rows = len(matrix)
        cols = len(matrix[0])
        first_row = any(matrix[0][c]==0 for c in range(cols))
        first_col = any(matrix[r][0]==0 for r in range(rows))

        # mark
        for r in range(1, rows):
            for c in range(1, cols):
                if matrix[r][c]==0:
                    matrix[0][c] = 0
                    matrix[r][0] = 0
        
        # zero based on marks
        for r in range(1, rows):
            for c in range(1, cols):
                if matrix[r][0]==0 or matrix[0][c]==0:
                    matrix[r][c] = 0

        # first row
        if first_row:
            for c in range(cols):
                matrix[0][c] = 0

        # first col
        if first_col:
            for r in range(rows):
                matrix[r][0] = 0